### PR TITLE
Change gas mixer to mix by molar instead of volumetric

### PR DESCRIPTION
**Pull request by Deadgebert**
## About The Pull Request

Gas mixer output now accurate to set ratio regardless of mismatched input temperatures
Input
53c841b2db6c6d5f0db15634c613c50c-1

Mixer
c619537ba1cdbfc5ec8a6c384aabc6c3

Output
2d9c662f348838f427c1523fd86c092b

## Why It's Good For The Game

 
We can start by looking at what gas mixers are actually good for. First we have air supply mix and burn mixes, these are the main uses for a mixer, but the resulting mix becomes inaccurate if one of your gas supply tanks is cooler or warmer than another (this often happens when someone adds a bunch of cold n2 or o2 to wastes). A gas mix for the HFR or crystallizer, usually these are set up with a constant supply of gasses with temperature of an input often varying, eventually clogging if the mix becomes inaccurate. Finally we have bomb and internal mixes, I would argue that this is the most important reason to keep things the way they are, but are actually currently the easiest to manage. All of these uses require molar precision, something the gas mixer doesn't really provide on its own, making it almost useless if two inputs do not match in temperature.

In my experience, in order to mix gasses accurately, you must do one of three things. First, you must use a combination of heat gates and thermomachines to equalize the temperatures before mixing. Second, you can combine the mixtures into one to equalize the temperature, filter the gasses apart, and mix again with your desired ratios. Or the third option, you could create a custom calculator to take in your desired ratio and the current temperatures of your input and plug the resulting ratio into the mixer. 

The way gas mixers work currently serves no real purpose outside of tripping up new players and providing 'busy work' for the experienced ones. Many newer players expect gas mixers to provide the set ratio to the output, not knowing that the temperatures on the input mixes must match. It requires a lot of additional effort and knowledge to accurately mix gasses, something that should just be provided by the tool itself. 

This change is good because it helps makes atmos a bit simpler and more intuitive. Unfortunately we will lose complexity in setups and make atmos a bit easier, but maybe these are good things

## Changelog

:cl:
balance: gas mixer output now accurate to set ratio regardless of input temperatures
/:cl:

### CODE REVIEW

**Comment by Acer202** [comment]
Make a switch to toggle it between pressure-based and mole-based perhaps?

**Comment by LemonInTheDark** [comment]
> Make a switch to toggle it between pressure-based and mole-based perhaps?

This is clutter. if this is better we should do it

**Comment by CRITAWAKETS** [comment]
I'm not sure why these even were the way that they were, if you're mixing gases with two different temperatures then they're going to have those two different temperatures mix together anyway.

File: code/modules/atmospherics/machinery/components/trinary_devices/mixer.dm [hunk 1]
```diff
@@ -84,8 +84,12 @@
...
 	var/general_transfer = (target_pressure - output_starting_pressure) * air3.volume / R_IDEAL_GAS_EQUATION
 
-	var/transfer_moles1 = air1.temperature ? node1_concentration * general_transfer / air1.temperature : 0
-	var/transfer_moles2 = air2.temperature ? node2_concentration * general_transfer / air2.temperature : 0
+	//Calculate combined temperature for accurate output ratio
+	var/combined_heat_capacity = air1.heat_capacity() + air2.heat_capacity()
+	var/equalized_temperature = combined_heat_capacity ? (air1.thermal_energy() + air2.thermal_energy()) / combined_heat_capacity : 0
+
+	var/transfer_moles1 = equalized_temperature ? node1_concentration * general_transfer / equalized_temperature : 0
+	var/transfer_moles2 = equalized_temperature ? node2_concentration * general_transfer / equalized_temperature : 0
```
Comment (san7890) [inline code/modules/atmospherics/machinery/components/trinary_devices/mixer.dm:92]:
```suggestion
	var/transfer_moles1 = equalized_temperature ? (node1_concentration * general_transfer) / equalized_temperature : 0
	var/transfer_moles2 = equalized_temperature ? (node2_concentration * general_transfer) / equalized_temperature : 0
```

i'm pretty sure BYOND will never devolve and process this the wrong way but i'm eternally fearful of the PEMDAS demon eating our liver

Comment (Deadgebert) [reply code/modules/atmospherics/machinery/components/trinary_devices/mixer.dm]: i agree, parens are good, even if just for readability. thanks for pointing it out. i've made the change and tested.

**Pull request merged**